### PR TITLE
Add unique_count algorithm

### LINKED ISSUE #1612: Add unique_count algorithm
For the common *count* -> allocate -> *fill* pattern (e.g. `count_if`/`copy_if`) there are some fill algorithms without a corresponding count algorithm, mainly `unique_copy`, `unique_by_key_copy` and `reduce_by_key`. So I want to suggest adding a `count_unique` algorithm with basically the same interface as `unique` save the return type:
```cpp
difference_type thrust::count_unique(
        Policy exec,
        ForwardIterator first,
        ForwardIterator last,
        BinaryPredicate binary_pred)
```

The name probably needs a bit of discussion, since it doesn't actually count unique elements, but runs of unique elements, but `unique` has the same issue, so there is some prior art for it.

@allisonvacanti suggested using [`cub::DeviceRunLengthEncode::Encode`](https://nvlabs.github.io/cub/structcub_1_1_device_run_length_encode.html) to implement it, so I wanted to check whether there are any other approaches to consider before starting work on a PR. Another simple approach might be
```cpp
auto zip_it = zip(it, it + 1);
if (size > 0) {
    return 1 + count_if(zip_it, zip_it + size - 1, [](auto a) { !binary_pred(get<0>(a), get<1>(a)); });
} else {
    return 0;
}
```

### PR DESCRIPTION
**Pull request by upsj**
Add a counting equivalent to unique_* algorithms that can be used to allocate the correct amount of data before actually filling it.

The formatting is a bit all over the place, I just tried to fit the environment as well as possible.

I'll run a few benchmarks to see how this fares performance-wise (I am hoping I won't need to use `cub::DeviceRunLengthEncode::Encode`, because that will be more work to integrate, and does more work internally)

I tested everything with the CUDA, OMP and CPU backends, don't have a TBB installation on hand, but the code is equivalent to OpenMP, so hopefully I didn't break anything :)

Closes #1612

### CODE REVIEW

**Comment by GPUtester** [comment]
Can one of the admins verify this patch?

File: thrust/system/cuda/detail/unique.h
Comment (upsj) [inline thrust/system/cuda/detail/unique.h:834]: What I haven't looked in deeper here: What if policy is a device policy, but first and last host pointers? I'm guessing that means that Thrust would create a temporary copy of the *tuples* on the device? That's probably not desirable

Comment (ericniebler) [reply thrust/system/cuda/detail/unique.h:834]: The answer is: don't do that. If you pass a device policy and host pointers, it blows up. Not supported.

File: thrust/system/cuda/detail/unique.h
Comment (ericniebler) [inline thrust/system/cuda/detail/unique.h:821]: Should be `ForwardIterator`, right?

File: thrust/system/cuda/detail/unique.h
Comment (ericniebler) [inline thrust/system/cuda/detail/unique.h:832]: This algorithm should work with forward iterators, but taking the difference of two iterators requires random access. You'll need to find a different way to implement this. @jrhemstad and @allisonvacanti suggest a possible implementation in #1619.

File: testing/unique.cu
Comment (ericniebler) [inline testing/unique.cu:338]: `unique_count` claims to work with forward iterators. There should be a test using forward iterators.

**Review by ericniebler — CHANGES_REQUESTED** [review]
Thanks for this. Needs a few tweaks.

File: thrust/system/cuda/detail/unique.h
Comment (upsj) [reply thrust/system/cuda/detail/unique.h:821]: Right, that was copy-pasted from the definition above. Should I change it there as well?

File: thrust/system/cuda/detail/unique.h
Comment (upsj) [reply thrust/system/cuda/detail/unique.h:832]: I suppose thrust::distance would be appropriate then? I am having a hard time imagining how you would implement this parallel algorithm without random access iterators.

Regardless of that, I'll give `DeviceRunLengthEncode::Encode` a try, since it's simpler than I thought on my first look.

File: testing/unique.cu
Comment (upsj) [reply testing/unique.cu:338]: Is there a nice way to wrap an iterator into a forward_iterator in Thrust? I wrote a small wrapper class and that seems to work and compile, but I suppose that problem has been solved elsewhere already?

File: thrust/system/cuda/detail/unique.h
Comment (upsj) [reply thrust/system/cuda/detail/unique.h:832]: Seems like cub doesn't like discard_iterator, since it writes and reads the first key from each run. So I think this is not possible without additional temporary storage O(size)

File: thrust/system/cuda/detail/unique.h
Comment (upsj) [reply thrust/system/cuda/detail/unique.h:834]: Thanks for the clarification, I thought I saw that working somewhere in the past

File: thrust/system/cuda/detail/unique.h
Comment (upsj) [inline thrust/system/cuda/detail/unique.h:834]: I'm not sure: Does zip_iterator inherit the forward_iterator category from `first`? If so, then we probably need something like
```suggestion
  return 1 + thrust::count_if(policy, it, thrust::advance(it, size - 1), zip_adj_not_predicate<BinaryPred>{binary_pred});
```

File: thrust/system/cuda/detail/unique.h
Comment (allisonvacanti) [reply thrust/system/cuda/detail/unique.h]: Yes, please do.

File: thrust/system/cuda/detail/unique.h
Comment (allisonvacanti) [reply thrust/system/cuda/detail/unique.h]: Using `thrust::distance` is correct here.

Could you file a CUB issue describing the discard iterator bug?

File: testing/unique.cu
Comment (allisonvacanti) [reply testing/unique.cu:338]: @ericniebler is right, we should be testing this. Unfortunately we lack robust testing for these sorts of things.

Thanks for adding the new testing infrastructure! Please include it in this PR, ideally in the testing framework so we can reuse it from other tests later 👍

File: thrust/system/cuda/detail/unique.h
Comment (allisonvacanti) [reply thrust/system/cuda/detail/unique.h:834]: Yes, please use `thrust::advance` instead. Note that it doesn't return a new iterator, but rather increments its argument, so the suggested fix will need to be refactored a bit.

Comment (upsj) [reply thrust/system/cuda/detail/unique.h:834]: right, I mixed up std::next and std::advance. Should Thrust have the former? Probably easy to add, too.

File: thrust/system/cuda/detail/unique.h
Comment (upsj) [reply thrust/system/cuda/detail/unique.h]: done

File: thrust/system/cuda/detail/unique.h
Comment (upsj) [reply thrust/system/cuda/detail/unique.h]: Turns out this has already been reported: https://github.com/NVIDIA/thrust/issues/1490

**Comment by jrhemstad** [comment]
> Is this kind of overhead acceptable?

I'm suspicious of why this would be slower than the `count_if` benchmark by a non-trivial amount. It should be moving the same amount of data from global memory. Technically each thread now loads 2 elements, but one of those is also loaded by the adjacent thread such that when threads load the first element it is coalesced and loading the second element it is coalesced and should easily hit in L1/L2. 

@upsj if you're interested, I'd be curious to see an Nsight Compute profile of those two kernels with emphasis on the memory movement summary. That would tell us if `unique_count` somehow has more global memory traffic.

File: thrust/system/cuda/detail/unique.h
Comment (upsj) [reply thrust/system/cuda/detail/unique.h]: Okay, `DeviceRunLengthEncode::Encode` works with `cub::DiscardOutputIterator`, but (at least on the Titan X I tested on) gives significantly worse performance, comparison (GB/s bandwidth)

int

| n | count_if | Encode |
|---|-------------|-----------|
| 16M |  179.173 | 109.111 |
| 1B | 254.523 | 160.037 |

long

| n | count_if | Encode |
|---|-------------|-----------|
| 16M |  259.123 | 180.335 |
| 512M | 315.969 | 215.304 |

pair<long, int>

| n | count_if | Encode |
|---|-------------|-----------|
| 16M |  351.678 284.543 |
| 256M | 392.726 | 331.611 |

**Comment by upsj** [comment]
This is definitely an issue specific to the implementation of count_if. With a quick-and-dirty hand-written implementation, I get performance on par with plain `count_if` for int and somewhere between count_if and unique_count for long.

<details>

```cpp
#include "thrust/unique.h"
#include "thrust/device_vector.h"
#include "thrust/iterator/zip_iterator.h"
#include "thrust/tuple.h"
#include "thrust/sequence.h"

#include <chrono>
#include <iostream>


template <typename T>
__global__ __launch_bounds__(1024) void unique_count_raw(const T* data, int size, int* out_counts) {
	const auto tidx = threadIdx.x + blockIdx.x * blockDim.x;
	const auto num_threads = gridDim.x * blockDim.x;
	__shared__ int counts[32];
	int count{};
	const auto lane = tidx % 32;
	for (auto i = tidx; i < size - 1; i += num_threads) {
		count += data[i] != data[i + 1];
	}
	#pragma unroll
	for (int step = 1; step < 32; step *= 2) {
		count += __shfl_xor_sync(0xFFFFFFFFu, count, step);
	}
	if (lane == 0) {
		counts[threadIdx.x / 32] = count;
	}
	__syncthreads();
	if (threadIdx.x >= 32) {
		return;
	}
	count = counts[lane];
	#pragma unroll
	for (int step = 1; step < 32; step *= 2) {
		count += __shfl_xor_sync(0xFFFFFFFFu, count, step);
	}
	if (lane == 0) {
		out_counts[blockIdx.x] = count;
	}
}


__device__ __host__ long get(int a) { return a; }
__device__ __host__ long get(long a) { return a; }
template <typename A, typename B>
__device__ __host__ long get(thrust::tuple<A,B> a) { return thrust::get<0>(a) + thrust::get<1>(a); }

template <typename ForwardIt>
void benchmark(std::string desc, ForwardIt begin, ForwardIt end) {
	thrust::unique_count(begin, end);
	cudaDeviceSynchronize();
	const auto reps = 100;
	auto start = std::chrono::high_resolution_clock::now();
	int count1;
	int count2;
	for (int i = 0; i < reps; i++) {
		count1 = thrust::unique_count(begin, end);
	}
	cudaDeviceSynchronize();
	auto total = std::chrono::duration_cast<std::chrono::nanoseconds>(std::chrono::high_resolution_clock::now() - start).count() / reps;
	cudaDeviceSynchronize();
	start = std::chrono::high_resolution_clock::now();
	for (int i = 0; i < reps; i++) {
		thrust::count_if(begin, end, []__host__ __device__(typename ForwardIt::value_type val) { return get(val) >= 0; });
	}
	cudaDeviceSynchronize();
	auto total2 = std::chrono::duration_cast<std::chrono::nanoseconds>(std::chrono::high_resolution_clock::now() - start).count() / reps;
	cudaDeviceSynchronize();
	start = std::chrono::high_resolution_clock::now();
	thrust::device_vector<int> tmp_storage(1024);
	for (int i = 0; i < reps; i++) {
		unique_count_raw<<<1024, 1024>>>((&begin[0]).get(), end - begin, tmp_storage.data().get());
		count2 = 1 + thrust::reduce(tmp_storage.begin(), tmp_storage.end());
	}
	cudaDeviceSynchronize();
	auto total3 = std::chrono::duration_cast<std::chrono::nanoseconds>(std::chrono::high_resolution_clock::now() - start).count() / reps;
	if (count1 != count2) {
		std::cerr << "Invalid: " << count1 << ',' << count2 << '\n';
	}
	auto memory = sizeof(typename ForwardIt::value_type) * (end - begin) * 1.0;
	std::cout << desc << ',' << (memory / total) << ',' << (memory / total2) << ',' << (memory / total3) << '\n';
}

int main() {
	{
		thrust::device_vector<int> data(1u << 30);
		thrust::sequence(data.begin(), data.end(), 0);
		for (long i = 1; i <= data.size(); i *= 2) {
			benchmark("int," + std::to_string(i), data.begin(), data.begin() + i);
		}
	}
	{
		thrust::device_vector<long> data(1u << 29);
		thrust::sequence(data.begin(), data.end(), 0);
		for (long i = 1; i <= data.size(); i *= 2) {
			benchmark("long," + std::to_string(i), data.begin(), data.begin() + i);
		}
	}
	/*{
		thrust::device_vector<int> data1(1u << 28);
		thrust::device_vector<long> data2(1u << 28);
		thrust::sequence(data1.begin(), data1.end(), 0);
		thrust::sequence(data2.begin(), data2.end(), 0);
		auto it = thrust::make_zip_iterator(thrust::make_tuple(data1.begin(), data2.begin()));
		for (long i = data1.size(); i <= data1.size(); i *= 2) {
			benchmark("pair<int long>," + std::to_string(i), it, it + i);
		}
	}*/
}


```

</details>

**Comment by upsj** [comment]
After taking a brief look over cub's Reduce implementation, I would guess that it is very well optimized for a particular setting, and that is pointer-like iterators that read exactly as much as their `value_type`. That is true for count_if, but not for count_unique. This imbalance, together with the fact that cub reads multiple entries in each thread with threads_per_block stride might lead to elements being evicted from L1 cache earlier than would be ideal?

File: testing/unittest/iterator_helpers.h
Comment (ericniebler) [inline testing/unittest/iterator_helpers.h:16]: Please also add a `static_assert` verifying that `Iterator` is _at least_ a forward iterator. Something like (untested):

```c++
using base_iterator_category = typename thrust::iterator_traits<Iterator>::iterator_category;
static_assert(
    std::is_convertible<base_iterator_category, std::forward_iterator_tag>::value, 
    "<insert a meaningful message goes here>");
```

**Review by ericniebler — CHANGES_REQUESTED** [review]
Thanks for adding the test! I have a few more suggested changes to the new `forward_iterator_wrapper` class, and a question for @allisonvacanti.

**Comment by upsj** [comment]
@ericniebler Thanks for the feedback, it's been a while since I've written truly generic code!

**Comment by ericniebler** [comment]
run tests

**Review by ericniebler — APPROVED** [review]

**Comment by jrhemstad** [comment]
@senior-zero any ideas about why this `unique_count` implementation is seeing 2x more global loads? 

I suspect something is going on with strided accesses as a result of the CUB reduction implementation.

File: testing/unique.cu
Comment (allisonvacanti) [reply testing/unique.cu:346]: This should be possible. Thrust packs some extra info into its iterator category. See `thrust/iterator/iterator_categories.h` for more info.

The iterator wrapper will need to test the base iterator using `thrust::iterator_system<BaseIterator>::type`, and define its iterator category to `forward_host_iterator_tag` if the result is `host_system_tag`, or `forward_device_iterator_tag` if the result is `device_iterator_tag`.

Comment (allisonvacanti) [reply testing/unique.cu:346]: This should be possible. Thrust packs some extra info into its iterator category. See `thrust/iterator/iterator_categories.h` for more info.

The iterator wrapper will need to test the base iterator using `thrust::iterator_system<BaseIterator>::type`, and define its iterator category to `forward_host_iterator_tag` if the result is `host_system_tag`, or `forward_device_iterator_tag` if the result is `device_iterator_tag`.

**Review by allisonvacanti — CHANGES_REQUESTED** [review]

File: testing/unique.cu
Comment (upsj) [reply testing/unique.cu:346]: I tried using this approach, but it looks like the implementation internally still tries to use the forward_iterator_wrapper like a random access iterator, so the compilation fails missing operator[] and operator+ and similar places.

Corresponding diff:
```diff
diff --git a/testing/unittest/iterator_helpers.h b/testing/unittest/iterator_helpers.h
index 8d4f03f5..8f254f46 100644
--- a/testing/unittest/iterator_helpers.h
+++ b/testing/unittest/iterator_helpers.h
@@ -1,6 +1,8 @@
 #pragma once
 
 #include <thrust/iterator/iterator_traits.h>
+#include <thrust/iterator/iterator_categories.h>
+#include <type_traits>
 
 
 // Wraps an existing iterator into a forward iterator,
@@ -8,11 +10,18 @@
 template <typename Iterator>
 struct forward_iterator_wrapper {
     // LegacyIterator requirements
+    using iterator_system_tag = typename thrust::iterator_system<Iterator>::type;
     using reference = typename thrust::iterator_traits<Iterator>::reference;
     using pointer = typename thrust::iterator_traits<Iterator>::pointer;
     using value_type = typename thrust::iterator_traits<Iterator>::value_type;
     using difference_type = typename thrust::iterator_traits<Iterator>::difference_type;
-    using iterator_category = std::forward_iterator_tag;
+    using iterator_category = typename std::conditional<
+        std::is_convertible<iterator_system_tag, thrust::device_system_tag>::value,
+        thrust::forward_device_iterator_tag,
+        typename std::conditional<
+            std::is_convertible<iterator_system_tag, thrust::host_system_tag>::value,
+            thrust::forward_host_iterator_tag,
+            std::forward_iterator_tag>::type>;
     using base_iterator_category = typename thrust::iterator_traits<Iterator>::iterator_category;
     static_assert(
         std::is_convertible<base_iterator_category, std::forward_iterator_tag>::value,
```

**Comment by senior-zero** [comment]
> @senior-zero any ideas about why this `unique_count` implementation is seeing 2x more global loads?
> 
> I suspect something is going on with strided accesses as a result of the CUB reduction implementation.

@jrhemstad I've profiled the latest version:
```cpp
    const int n = 128 * 1024 * 1024;
    thrust::device_vector<std::size_t> a(n);
    thrust::sequence(a.begin(), a.end());
    thrust::unique_count(a.begin(), a.end());
```
Here are the results:
|   |  expected | actual |
|---|---|---|
| int  |  0.5 GB |  0.56 GB  |
|  size_t |  1 GB | 1.09 GB |

Am I missing the reproducer that leads to 2x global loads?

**Comment by jrhemstad** [comment]
@senior-zero good question! I haven't tried running it myself. I was going off of @upsj's comment here https://github.com/NVIDIA/thrust/pull/1619#issuecomment-1065328895

@upsj are you not seeing the perf regression anymore?

**Comment by upsj** [comment]
@senior-zero @jrhemstad I was talking about load instruction executed, not amount of memory served by DRAM (i.e. `gld_transactions`, not `dram_read_transactions` from `nvprof` output) . The 2x amount of global memory loads there is to be expected, because we always need to load two adjacent elements per thread. Let me visualize the issue some more with some NSight Compute results: 
Screenshot 2022-04-17 230656
I'm comparing the DeviceReduceKernel from `unique_count` with the hand-written `unique_count_raw` from https://github.com/NVIDIA/thrust/pull/1619#issuecomment-1065359919, the rest is your sample code with size 512M instead of 128M. 
We have a noticeably worse L1 hit rate and a higher total DRAM access volume compared to a simple hand-written implementation. For the reasons, I speculated in https://github.com/NVIDIA/thrust/pull/1619#issuecomment-1066141873

**Comment by ericniebler** [comment]
run tests

**Comment by allisonvacanti** [comment]
run tests

File: testing/unique.cu
Comment (allisonvacanti) [reply testing/unique.cu:346]: @upsj What's the current status of the dispatch issues? You'd mentioned that this was doing odd things like dispatching host_vectors to device and vice-versa -- was that resolved?

Comment (upsj) [reply testing/unique.cu:346]: Let me take another look at it, I haven't touched this in a while

Comment (upsj) [reply testing/unique.cu:346]: @allisonvacanti forgot a ::type in my std::conditional, it seems to work now :)

Comment (allisonvacanti) [reply testing/unique.cu:346]: This sounded odd to me, as I've never seen any logic in Thrust that would dispatch forward iterators to serial implementations. So I dug into it, and unfortunately this is due to a pretty nasty bug in Thrust's iterator traits.

The details are gory, but I've summarized them [in a comment](https://github.com/NVIDIA/thrust/issues/902#issuecomment-1119107644) on #902.

Until that's fixed, I'm not comfortable merging this test using the forward_iterator_wrapper, since they only "do the right thing" because the iterator framework is broken.

I hate to churn on this PR even more, but I think we should remove the iterator wrappers for now and just test that the regular iterators work. We can re-introduce the wrapper tests as part of #55, after #902 is fixed and settled.

@ericniebler Can you review the two linked issues and see if you agree with my suggestion?

**Review by allisonvacanti — CHANGES_REQUESTED** [review]
Requesting changes to block merging until the situation is https://github.com/NVIDIA/thrust/pull/1619/files#r866368806 is resolved.

File: testing/unique.cu
Comment (allisonvacanti) [reply testing/unique.cu:346]: I don't think that forward iterators actually need to be dispatched to the sequential backend. They support multipass reading and should be usable in a parallel algorithm, so long as they're only copied and incremented. Is there something in the unique_count/count_if algorithms that would break them?

Comment (upsj) [reply testing/unique.cu:346]: Oof, that sounds annoying. The main issue I see with parallel execution on forward iterators is that they introduce an essentially linear dependency chain that means either every thread ´i` starts from `begin` and increments it `n` times, or waits until one of its predecessors `j` is done and writes its iterator somewhere, to then increments it `i - j` times. Both don't really seem useful for parallel execution to me.

**Comment by upsj** [comment]
@allisonvacanti I reverted the changes to still have them available in case you need them later. Does that work for you?

**Review by jrhemstad — APPROVED** [review]

File: testing/unique.cu
Comment (allisonvacanti) [reply testing/unique.cu:338]: It will require more increments, but if the work-per-element is expensive compared to the cost of the iterator increment, it may still make sense to parallelize. I'd rather let the user make that call, since they can opt-in to sequential execution by passing in the sequential exeuction policy (`thrust::seq`).

More importantly, the sequential implementation executes on CPU, and some types of device memory aren't accessible from the CPU's address space, so switching to `seq` really needs to be opt-in rather than default.

Comment (upsj) [reply testing/unique.cu:338]: Thanks for the clarification, that makes sense! I was only thinking of simple but massively parallel cases.

**Comment by allisonvacanti** [comment]
Tests look good, merging. Thanks for the PR!

**Pull request merged**
